File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/GameViewPresenterImpl.cpp
#include "GameViewPresenterImpl.h"

#include "Territory.h"
#include "GameTerritory.h"
#include "GameLog.h"
#include "Semaphore.h"
#include "GameCommandStack.h"

#include "ViewModelCellLayer.h"

#include "CollectionHelpers.hpp"
#include <stdexcept>
#include <utility>

using namespace mpw;
using namespace hamster;
using namespace collectionhelpers;
using namespace framework;

namespace hamsterviewmodel {

// TODO remove listeners on removement of relating objects

GameViewPresenterImpl::GameViewPresenterImpl(std::shared_ptr<hamster::HamsterGame> game)
    : inherited(Semaphore())
    , game(std::move(game)) {
    const Size& size = this->game->getTerritory()->getTerritorySize();
    getViewModel()->init(size);
}

void GameViewPresenterImpl::bind() {
    auto lock = getSemaphore().lock();

    ReadOnlyTerritory& territory = *game->getTerritory()->getInternalTerritory();
    auto& tilesProperty = territory.tilesProperty();
    
    tilesProperty.addOnAddedListener([this](auto& tile) { addTileNode(tile); });
    tilesProperty.addOnRemovedListener([this](auto& tile) { removeTileNode(tile); });
    tilesProperty.forEach([this](auto& tile) { addTileNode(tile); });

    auto gameLog = game->getGameLog();
    gameLog->logEntriesProperty().addOnAddedListener([this](const std::string& entry) {
        auto lock = getSemaphore().lock();
        addLogEntry(entry);
    });
    gameLog->logEntriesProperty().forEach([this](const std::string& entry) {
        addLogEntry(entry);
    });
}

void GameViewPresenterImpl::playClicked() {
    game->getGameCommandStack()->resume();
}

void GameViewPresenterImpl::pauseClicked() {
    game->getGameCommandStack()->pause();
}

void GameViewPresenterImpl::undoClicked() {
    game->getGameCommandStack()->undo();
}

void GameViewPresenterImpl::redoClicked() {
    game->getGameCommandStack()->redo();
}

void GameViewPresenterImpl::textTyped(std::string text) {
    throw std::runtime_error("not implemented");
}

void GameViewPresenterImpl::addTileNode(const mpw::Tile& tile) {
    Tile& nonConstTile = const_cast<Tile&>(tile); // TODO const-correctness: adapt after migrate generator
    const Location& location = nonConstTile.getLocation();

    auto& contentsProperty = nonConstTile.contentsProperty();
    addedContentListenerIds[&tile] = contentsProperty.addOnAddedListener([this, &nonConstTile](const mpw::TileContent&) {
        auto lock = getSemaphore().lock();
        setTileNodeAt(nonConstTile.getLocation(), nonConstTile);
    });
    removedContentListenerIds[&tile] = contentsProperty.addOnRemovedListener([this, &nonConstTile](const mpw::TileContent&) {
        auto lock = getSemaphore().lock();
        setTileNodeAt(nonConstTile.getLocation(), nonConstTile);
    });
    setTileNodeAt(location, tile);
}

void GameViewPresenterImpl::removeTileNode(const mpw::Tile& tile) {
    Tile& nonConstTile = const_cast<Tile&>(tile); // TODO const-correctness: adapt after migrate generator
    auto cell = getViewModel()->getCellAt(
            nonConstTile.getLocation().getRow(),
            nonConstTile.getLocation().getColumn());
    cell->clearLayers();
}

void GameViewPresenterImpl::setTileNodeAt(const Location& location, const Tile& tile) {
    auto cell = getViewModel()->getCellAt(location.getRow(), location.getColumn());
    cell->clearLayers();

    configureWallImageView(*cell, tile);
    configureGrainImageView(*cell, tile);
    for (std::shared_ptr<ReadOnlyHamster> hamster : getHamstersOfTile(tile)) {
        configureHamsterImageView(*cell, *hamster);
    }

}

void GameViewPresenterImpl::configureGrainImageView(ViewModelCell& cell, const Tile& tile) {
    Tile& nonConstTile = const_cast<Tile&>(tile); // TODO const-correctness: adapt after migrate generator
    auto grainLayer = std::make_shared<ViewModelCellLayer>();
    refreshGrainLayer(*grainLayer, tile);

    cell.addToLayers(grainLayer);
}

void GameViewPresenterImpl::refreshGrainLayer(ViewModelCellLayer& layer, const mpw::Tile& tile) {
    size_t grainCount = getGrainOfTile(tile).size();
    layer.setVisible(grainCount > 0);
    if (grainCount <= 12) {
        layer.setImageName(std::to_string(grainCount) + "Corn32");
    } else {
        layer.setImageName("12PlusCorn32");
    }
}

void GameViewPresenterImpl::configureWallImageView(ViewModelCell& cell, const Tile& tile) {
    Tile& nonConstTile = const_cast<Tile&>(tile); // TODO const-correctness: adapt after migrate generator
    auto wallLayer = std::make_shared<ViewModelCellLayer>();
    wallLayer->setImageName("Wall32");
    refreshWallLayer(*wallLayer, tile);

    cell.addToLayers(wallLayer);
}

void GameViewPresenterImpl::refreshWallLayer(ViewModelCellLayer& layer, const Tile& tile) {
    layer.setVisible(!getWallsOfTile(tile).empty());
}

void GameViewPresenterImpl::configureHamsterImageView(ViewModelCell& cell, const ReadOnlyHamster& hamster) {
    auto& nonConstHamster = const_cast<ReadOnlyHamster&>(hamster); // TODO const-correctness: adapt after migrate generator
    auto hamsterLayer = std::make_shared<ViewModelCellLayer>();
    hamsterLayer->setImageName("Hamster32");

    changedHamsterDirectionListenerIds[&hamster] = nonConstHamster.directionProperty().addListener([this, &nonConstHamster, hamsterLayer](Direction oldValue, Direction newValue) {
        auto lock = getSemaphore().lock();
        refreshHamsterLayer(*hamsterLayer, nonConstHamster);
    });
    refreshHamsterLayer(*hamsterLayer, nonConstHamster);

    cell.addToLayers(hamsterLayer);
}

void GameViewPresenterImpl::refreshHamsterLayer(ViewModelCellLayer& layer, hamster::ReadOnlyHamster& hamster) {
    layer.setVisible(hamster.getCurrentTile() != nullptr);
    if (hamster.getCurrentTile() == nullptr) {
        return;
    }
    layer.setRotation(getRotationForDirection(hamster.getDirection()));
}

std::list<std::shared_ptr<hamster::ReadOnlyHamster>> GameViewPresenterImpl::getHamstersOfTile(const Tile& tile) {
    return type_select<hamster::ReadOnlyHamster>(const_cast<Tile&>(tile).getContents());
}

std::list<std::shared_ptr<hamster::Wall>> GameViewPresenterImpl::getWallsOfTile(const Tile& tile) {
    return type_select<hamster::Wall>(const_cast<Tile&>(tile).getContents());
}

std::list<std::shared_ptr<hamster::Grain>> GameViewPresenterImpl::getGrainOfTile(const Tile& tile) {
    return type_select<hamster::Grain>(const_cast<Tile&>(tile).getContents());
}

int GameViewPresenterImpl::getRotationForDirection(mpw::Direction direction) {
    switch (direction) {
        case Direction::EAST:
            return 0;
        case Direction::SOUTH:
            return 90;
        case Direction::WEST:
            return 180;
        case Direction::NORTH:
            return 270;
    }
    throw std::runtime_error("Invalid direction!");
}

void GameViewPresenterImpl::addLogEntry(const std::string& entry) {
    getViewModel()->addToLogEntries(entry);
}


}

File: de.unistuttgart.iste.sqa.mpw.karasimulator.java/core/src/main/java/de/unistuttgart/kara/hamsterviewmodel/impl/GameViewPresenterImpl.java
package de.unistuttgart.kara.hamsterviewmodel.impl;

import de.unistuttgart.kara.hamsterviewmodel.GameViewPresenter;
import de.unistuttgart.kara.hamsterviewmodel.ViewModelCell;
import de.unistuttgart.kara.hamsterviewmodel.ViewModelCellLayer;
import de.unistuttgart.kara.kara.*;
import de.unistuttgart.kara.mpw.*;
import javafx.beans.InvalidationListener;
import javafx.collections.ListChangeListener;

import java.util.Optional;
import java.util.concurrent.Semaphore;

public class GameViewPresenterImpl extends GameViewPresenter {
	private final KaraGame game;

	public GameViewPresenterImpl(KaraGame game) {
		super(new Semaphore(1, true));
		this.game = game;

		final Size size = this.game.getWorld().getWorldSize();
		getViewModel().init(size);
	}

	@Override
	public void bind() {
		runBlocking(() -> {
			var world = game.getWorld().getInternalWorld();
			world.tilesProperty().addListener(tilesChangedListener);
			world.tilesProperty().forEach(tile -> addTileNode(tile));

			var gameLog = game.getGameLog();
			gameLog.logEntriesProperty().addListener(logChangedListener);
			gameLog.logEntriesProperty().forEach(entry -> addLogEntry(entry));
		});
	}

	/**
	 * Log Handling
	 */

	private final ListChangeListener<String> logChangedListener = new ListChangeListener<>() {
		@Override
		public void onChanged(final Change<? extends String> change) {
			runBlocking(() -> {
				while (change.next()) {
					if (change.wasAdded()) {
						change.getAddedSubList().forEach(entry -> addLogEntry(entry));
					}
					if (change.wasRemoved()) {
						change.getRemoved().forEach(entry -> removeLogEntry(entry));
					}
				}
			});
		}
	};

	private void addLogEntry(String entry) {
		getViewModel().addToLogEntries(entry);
	}

	private void removeLogEntry(String entry) {
		getViewModel().removeFromLogEntries(entry);
	}

	/**
	 * Tile Handling
	 */

	private final ListChangeListener<Tile> tilesChangedListener = new ListChangeListener<Tile>() {
		@Override
		public void onChanged(final Change<? extends Tile> change) {
			runBlocking(() -> {
				while (change.next()) {
					if (change.wasAdded()) {
						change.getAddedSubList().forEach(tile -> addTileNode(tile));
					}
					if (change.wasRemoved()) {
						change.getRemoved().forEach(tile -> removeTileNode(tile));
					}
				}
			});
		}
	};

	private void addTileNode(final Tile tile) {
		final Location location = tile.getLocation();
		tile.contentsProperty().addListener((v, c, l) -> {
			runBlocking(() -> {
				setTileNodeAt(tile.getLocation(), tile);
			});
		});
		setTileNodeAt(location, tile);
	}

	private void removeTileNode(final Tile tile) {
		var cell = getViewModel().getCellAt(tile.getLocation().getRow(), tile.getLocation().getColumn());
		cell.getLayers().clear();
	}

	private void setTileNodeAt(final Location location, final Tile tile) {
		var cell = getViewModel().getCellAt(location.getRow(), location.getColumn());
		cell.getLayers().clear();

		configureTreeImageView(cell, tile);
		configureLeafImageView(cell, tile);
		configureMushroomImageView(cell, tile);

		Optional<ReadOnlyKara> karaOptional = findKaraOnTile(tile);
		karaOptional.ifPresent(readOnlyKara -> configureKaraImageView(cell, readOnlyKara));
	}

	private Optional<ReadOnlyKara> findKaraOnTile(Tile tile) {
		return tile.getContents().stream()
				.filter(ReadOnlyKara.class::isInstance)
				.map(ReadOnlyKara.class::cast).findFirst();
	}

	private void configureTreeImageView(ViewModelCell cell, Tile tile) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Tree32");
		refreshTreeLayer(layer, tile);
		cell.getLayers().add(layer);
	}

	private void refreshTreeLayer(ViewModelCellLayer layer, Tile tile) {
		layer.setVisible(tile.getContents().stream().anyMatch(Tree.class::isInstance));
	}

	private void configureLeafImageView(ViewModelCell cell, Tile tile) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Leaf32");
		refreshLeafLayer(layer, tile);
		cell.getLayers().add(layer);
	}

	private void refreshLeafLayer(ViewModelCellLayer layer, Tile tile) {
		layer.setVisible(tile.getContents().stream().anyMatch(Leaf.class::isInstance));
	}

	private void configureMushroomImageView(ViewModelCell cell, Tile tile) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Mushroom32");
		refreshMushroomLayer(layer, tile);
		cell.getLayers().add(layer);
	}

	private void refreshMushroomLayer(ViewModelCellLayer layer, Tile tile) {
		layer.setVisible(tile.getContents().stream().anyMatch(Mushroom.class::isInstance));
	}

	private void configureKaraImageView(ViewModelCell cell, ReadOnlyKara kara) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Kara32");

		kara.directionProperty().addListener((v, c, l) -> {
			runBlocking(() -> {
				refreshKaraLayer(layer, kara);
			});
		});

		refreshKaraLayer(layer, kara);
		cell.getLayers().add(layer);
	}

	private void refreshKaraLayer(ViewModelCellLayer layer, ReadOnlyKara kara) {
		layer.setVisible(kara.getCurrentTile() != null);
		layer.setRotation(getRotationForDirection(kara.getDirection()));
	}

	private int getRotationForDirection(final Direction direction) {
		assert direction != null;

		switch (direction) {
			case EAST:
				return 0;
			case SOUTH:
				return 90;
			case WEST:
				return 180;
			case NORTH:
				return 270;
		}
		throw new RuntimeException("Invalid direction!");
	}

	@Override
	public void playClicked() {
		game.getGameCommandStack().resume();
	}

	@Override
	public void pauseClicked() {
		game.getGameCommandStack().pause();
	}

	@Override
	public void undoClicked() {
		game.getGameCommandStack().undo();
	}

	@Override
	public void redoClicked() {
		game.getGameCommandStack().redo();
	}

	@Override
	public void textTyped(String text) {
		throw new RuntimeException("not implemented");
	}

	private void runBlocking(Runnable r) {
		try {
			getSemaphore().acquire();
			r.run();
		} catch (final Exception e) {
			throw new RuntimeException(e);
		} finally {
			getSemaphore().release();
		}
	}

}
